core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala

```diff
@@ -565,12 +603,25 @@ class PairRDDFunctions[K, V](self: RDD[(K, V)])
   }
 
   /**
-   * Simplified version of combineByKey that hash-partitions the resulting RDD using the
-   * existing partitioner/parallelism level.
+   * This method is here for backward compatibility. It
+   * does not provide combiner classtag information to
+   * the shuffle.
+   *
+   * @see [[combineByKeyWithClassTag]]
    */
   def combineByKey[C](createCombiner: V => C, mergeValue: (C, V) => C, mergeCombiners: (C, C) => C)
-    : RDD[(K, C)] = self.withScope {
-    combineByKey(createCombiner, mergeValue, mergeCombiners, defaultPartitioner(self))
+    : RDD[(K, C)] = {
+    combineByKeyWithClassTag(createCombiner, mergeValue, mergeCombiners)(null)
+  }
+
+  /**
+   * Simplified version of combineByKeyWithClassTag that hash-partitions the resulting RDD using the
+   * existing partitioner/parallelism level.
+   */
+  def combineByKeyWithClassTag[C](createCombiner: V => C, mergeValue: (C, V) => C,
+                                  mergeCombiners: (C, C) => C)
+    (implicit ct: ClassTag[C]): RDD[(K, C)] = self.withScope {
```

Comment:
The original `combineByKey` methods had `self.withScope` so the new `combineByKeyWithClassTags` should too, I believe.

I was worried about nesting the scopes but I think that may be beneficial (and I'll take your advice to re-add them). We want users to be able use the UI to debug if their application is using the `combineByKey` or the `combineByKeyWithClassTags` methods, since some shuffle implementation will need `ClassTag` information.